src/ring_theory/dedekind_domain.lean

```diff
@@ -0,0 +1,137 @@
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Kenji Nakagawa, Anne Baanen, Filippo A. E. Nuccio
+-/
+import ring_theory.discrete_valuation_ring
+import ring_theory.fractional_ideal
+import ring_theory.ideal.over
+
+/-!
+# Dedekind domains
+
+This file defines the notion of a Dedekind domain (or Dedekind ring),
+giving three equivalent definitions (TODO: and shows that they are equivalent).
+
+## Main definitions
+
+ - `is_dedekind_domain` defines a Dedekind domain as a commutative ring that is not a field,
+   Noetherian, integrally closed in its field of fractions and has Krull dimension exactly one.
+   `is_dedekind_domain_iff` shows that this does not depend on the choice of field of fractions.
+ - `is_dedekind_domain_dvr` alternatively defines a Dedekind domain as an integral domain that is not a field,
+   Noetherian, and the localization at every nonzero prime ideal is a discrete valuation ring.
+ - `is_dedekind_domain_inv` alternatively defines a Dedekind domain as an integral domain that is not a field,
+   and every nonzero fractional ideal is invertible.
+ - `is_dedekind_domain_inv_iff` shows that this does note depend on the choice of field of fractions.
+
+## Implementation notes
+
+The definitions that involve a field of fractions choose a canonical field of fractions,
+but are independent of that choice. The `..._iff` lemmas express this independence.
+
+## References
+
+* [D. Marcus, *Number Fields*][marcus1977number]
+* [J.W.S. Cassels, A. Frölich, *Algebraic Number Theory*][cassels1967algebraic]
+
+## Tags
+
+dedekind domain, dedekind ring
+-/
+
+variables (R A K : Type*) [comm_ring R] [integral_domain A] [field K]
+
+/-- A ring `R` has Krull dimension at most one if all nonzero prime ideals are maximal. -/
+def ring.dimension_le_one : Prop :=
+∀ p ≠ (⊥ : ideal R), p.is_prime → p.is_maximal
+
+open ideal ring
+
+lemma is_principal_ideal_ring.dimension_le_one
+  [is_principal_ideal_ring A] : ring.dimension_le_one A :=
+λ p nonzero prime, by { haveI := prime, exact is_prime.to_maximal_ideal nonzero }
+
+lemma integral_closure.dimension_le_one [nontrivial R] [algebra R A]
+  (h : dimension_le_one R) : dimension_le_one (integral_closure R A) :=
+begin
+  intros p ne_bot prime,
+  haveI := prime,
+  refine integral_closure.is_maximal_of_is_maximal_comap p
+    (h _ (integral_closure.comap_ne_bot ne_bot) _),
+  apply is_prime.comap
+end
+
+/--
+A Dedekind domain is a nonfield that is Noetherian, integrally closed, and
+has Krull dimension exactly one (`not_is_field` and `dimension_le_one`).
+
+The integral closure condition is independent of the choice of field of fractions:
+use `is_dedekind_domain_iff` to prove `is_dedekind_domain` for a given `fraction_map`.
+
+This is the default implementation, but there are equivalent definitions,
+`is_dedekind_domain_dvr` and `is_dedekind_domain_inv`.
+TODO: Prove that these are actually equivalent definitions.
+-/
+-- TODO: R should be a `comm_ring`
```

Comment:
Why do you say this? I think that R should be an `integral_domain` -- I thought we had established that `ring` and `comm_ring` and `integral_domain` should be the three "base classes" with data, and that everything else should build on them.

Reply:
Indeed, this TODO was outdated, and it should have been removed.